README.md
# godice-rolls
FoundryVTT module to use goDice bluetooth dices 


godice-rolls/CHANGELOG.md
# goDice Rolls

## Release 1.0.0 (2022-08-06)

- Official Release


scripts/godice.js

/**
 * @class
 * The main GoDice class that can be used to connect a new die, send and recieve messages.
 */
class GoDice {

	messageIdentifiers = {
		BATTERY_LEVEL: 3,
		DICE_COLOUR: 23,
		SET_LED: 8,
		SET_LED_TOGGLE: 16,
	}
	
	diceColour = {
		BLACK: 0,
		RED: 1,
		GREEN: 2,
		BLUE: 3,
		YELLOW: 4,
		ORANGE: 5,
	}
	
	static diceTypes = {
		D6: 0,
		D20: 1,
		D10: 2,
		D10X: 3,
		D4: 4,
		D8: 5,
		D12: 6,
	}
	
	// The vectors for each die and shell type, used to check which side is facing up
	static d6Vectors = {
		1: [-64, 0, 0],
		2: [0, 0, 64],
		3: [0, 64, 0],
		4: [0, -64, 0],
		5: [0, 0, -64],
		6: [64, 0, 0],
	}
	
	static d20Vectors = {
		1: [-64, 0, -22],
		2: [42, -42, 40],
		3: [0, 22, -64],
		4: [0, 22, 64],
		5: [-42, -42, 42],
		6: [22, 64, 0],
		7: [-42, -42, -42],
		8: [64, 0, -22],
		9: [-22, 64, 0],
		10: [42, -42, -42],
		11: [-42, 42, 42],
		12: [22, -64, 0],
		13: [-64, 0, 22],
		14: [42, 42, 42],
		15: [-22, -64, 0],
		16: [42, 42, -42],
		17: [0, -22, -64],
		18: [0, -22, 64],
		19: [-42, 42, -42],
		20: [64, 0, 22],
	}
	
	static d24Vectors = {
		1: [20, -60, -20],
		2: [20, 0, 60],
		3: [-40, -40, 40],
		4: [-60, 0, 20],
		5: [40, 20, 40],
		6: [-20, -60, -20],
		7: [20, 60, 20],
		8: [-40, 20, -40],
		9: [-40, 40, 40],
		10: [-20, 0, 60],
		11: [-20, -60, 20],
		12: [60, 0, 20],
		13: [-60, 0, -20],
		14: [20, 60, -20],
		15: [20, 0, -60],
		16: [40, -20, -40],
		17: [-20, 60, -20],
		18: [-40, -40, -40],
		19: [40, -20, 40],
		20: [20, -60, 20],
		21: [60, 0, -20],
		22: [40, 20, -40],
		23: [-20, 0, -60],
		24: [-20, 60, 20],
	}
	
	// Transforms from each shell type to according number on shell
	// D20 Transforms
	static d10Transform = {
		1: 8,
		2: 2,
		3: 6,
		4: 1,
		5: 4,
		6: 3,
		7: 9,
		8: 0,
		9: 7,
		10: 5,
		11: 5,
		12: 7,
		13: 0,
		14: 9,
		15: 3,
		16: 4,
		17: 1,
		18: 6,
		19: 2,
		20: 8,
	}
	
	static d10XTransform = {
		1: 80,
		2: 20,
		3: 60,
		4: 10,
		5: 40,
		6: 30,
		7: 90,
		8: 0,
		9: 70,
		10: 50,
		11: 50,
		12: 70,
		13: 0,
		14: 90,
		15: 30,
		16: 40,
		17: 10,
		18: 60,
		19: 20,
		20: 80,
	}
	
	// D24 Transforms
	static d4Transform = {
		1: 3,
		2: 1,
		3: 4,
		4: 1,
		5: 4,
		6: 4,
		7: 1,
		8: 4,
		9: 2,
		10: 3,
		11: 1,
		12: 1,
		13: 1,
		14: 4,
		15: 2,
		16: 3,
		17: 3,
		18: 2,
		19: 2,
		20: 2,
		21: 4,
		22: 1,
		23: 3,
		24: 2,
	}
	
	static d8Transform = {
		1: 3,
		2: 3,
		3: 6,
		4: 1,
		5: 2,
		6: 8,
		7: 1,
		8: 1,
		9: 4,
		10: 7,
		11: 5,
		12: 5,
		13: 4,
		14: 4,
		15: 2,
		16: 5,
		17: 7,
		18: 7,
		19: 8,
		20: 2,
		21: 8,
		22: 3,
		23: 6,
		24: 6,
	}
	
	static d12Transform = {
		1: 1,
		2: 2,
		3: 3,
		4: 4,
		5: 5,
		6: 6,
		7: 7,
		8: 8,
		9: 9,
		10: 10,
		11: 11,
		12: 12,
		13: 1,
		14: 2,
		15: 3,
		16: 4,
		17: 5,
		18: 6,
		19: 7,
		20: 8,
		21: 9,
		22: 10,
		23: 11,
		24: 12,
	}
	
	bluetoothDevice;
	GoDiceService;
	CubeCharacteristics;
	GlobalDeviceId;
	diceId;
	rolledValue = 0;
	dieType = GoDice.diceTypes.D6; // Type of die
	dieColor = this.diceColour.BLACK;
	dieBatteryLevel = 0;
	newConnection = true;

	onRollStart(){};
	onBatteryLevel(){};
	onDiceColor(){};
	onStable(){};
	onFakeStable(){};
	onTiltStable(){};
	onMoveStable(){};
	onDiceConnected(){};

	/******* API functions *******/

	constructor(diceInstance)
	{
		if(diceInstance != null)
		{
			this.GlobalDeviceId = diceInstance.GlobalDeviceId;
			this.dieType = diceInstance.dieType;
			this.dieColor = diceInstance.dieColor;
			this.newConnection = false;
		}
	}

	/**
	 * Request for the die battery, that should follow by corresponding "BatteryLevel" event (response).
	 */
	getBatteryLevel() {
		console.log(this)
		this.sendMessage([this.messageIdentifiers.BATTERY_LEVEL]);
	}

	/**
	 * Retunr stored die color or request and wait for the die color (that should follow by corresponding "DiceColor" event (response)).
	 */
	setDiceColor() {
		console.log(this);
		//this.sendMessage([this.messageIdentifiers.DICE_COLOUR]);
		if(this.bluetoothDevice.name.includes("_K_"))
			this.dieColor = this.diceColour.BLACK;
		else if(this.bluetoothDevice.name.includes("_R_"))
			this.dieColor = this.diceColour.RED;
		else if(this.bluetoothDevice.name.includes("_G_"))
			this.dieColor = this.diceColour.GREEN;
		else if (this.bluetoothDevice.name.includes("_B_"))
			this.dieColor = this.diceColour.BLUE;
		else if(this.bluetoothDevice.name.includes("_Y_"))
			this.dieColor =  this.diceColour.YELLOW;
		else if(this.bluetoothDevice.name.includes("_O_"))
			this.dieColor = this.diceColour.ORANGE;
		else 
			console.log("Dice Color not handled ornot present in the name, please use retrieveDiceColor() async to retrieve the value");
	}

	getDiceColor(){
		return dieColor;
	}

	retrieveDiceColor()
	{
		console.log(this);
		//this.sendMessage([this.messageIdentifiers.DICE_COLOUR]);
	}

	setDieType(newDieType) {
		this.dieType = newDieType
	}
	
	//return stored dice type
	getDieType() {
		return this.dieType
	}
	
	/**
	 * Open a connection dialog to connect a single GoDice, after successfull connection it will follow by corresponding "DiceConnected" event (response).
	 */
	requestDevice() {
		return navigator.bluetooth.requestDevice({
			filters: [{ namePrefix: 'GoDice_' }],
			optionalServices: ['6e400001-b5a3-f393-e0a9-e50e24dcca9e']
		})
			.then(device => {				
				this.GlobalDeviceId = device.id.toString();				
				this.bluetoothDevice = device;				
				this.bluetoothDevice.addEventListener('gattserverdisconnected', this.onDisconnected);				
				this.connectDeviceAndCacheCharacteristics();				
			});
	}

	/**
	 * Open a connection dialog to connect a single GoDice, after successfull connection it will follow by corresponding "DiceConnected" event (response).
	 */
	reconnectDevice() {
		if (!this.GlobalDeviceId) {
			return Promise.reject(new Error('No device ID available, use requestDevice instead.'));
		}
		return navigator.bluetooth.getDevices()
			.then(devices => {	
				for(const device of devices)
				{
					if(device.id == this.GlobalDeviceId)
					{
						this.GlobalDeviceId = device.id.toString();				
						this.bluetoothDevice = device;
						this .newConnection = false;				
						this.bluetoothDevice.addEventListener('gattserverdisconnected', this.onDisconnected);				
						this.reConnectDeviceAndCacheCharacteristics();
					}	
				}			
			});
	}

	/**
	 * Turn On/Off RGB LEDs, will turn off if led1 and led2 are null
	 * @param {Array} led1 - an array to control the 1st LED in the following format '[R, G, B]'
	 *                       where R,G and B are numbers in the range of 0-255
	 *
	 * @param {Array} led2 - an array to control the 2nd LED in the following format '[R, G, B]'
	 *                       where R,G and B are numbers in the ran	ge of 0-255
	 */
	setLed(led1, led2) {
		console.log(led1, led2)
		let adjustedLed1 = (!led1) ? [0, 0, 0] : led1;
		let adjustedLed2 = (!led2) ? [0, 0, 0] : led2;
		adjustedLed1 = adjustedLed1.map((i) => Math.max(Math.min(i, 255), 0));
		adjustedLed2 = adjustedLed2.map((i) => Math.max(Math.min(i, 255), 0));
		if (adjustedLed1.length === 1) adjustedLed1.push(adjustedLed1[0], adjustedLed1[0]);
		if (adjustedLed2.length === 1) adjustedLed2.push(adjustedLed2[0], adjustedLed2[0]);

		const messageArray = [this.messageIdentifiers.SET_LED, ...adjustedLed1, ...adjustedLed2];
		console.debug("messageArray", messageArray);
		this.sendMessage(messageArray);
	}
	
	/**
	 * Pulses LEDs for set time and color
	 * @param {number} pulseCount - an integer of how many times the pulse will repeat (max 255)
	 * @param {number} onTime 	- how much time should the LED be ON each pulse (units of 10ms, max 255) 
	 * @param {number} offTime 	- how much time should the LED be OFF each pulse (units of 10ms, max 255)
	 * @param {Array}  RGB  - an array to control both LEDs color's in the following format '[R, G, B]' 
	 * 						 where R, G and B are number in the range of 0-255
	 */
	pulseLed(pulseCount, onTime, offTime, RGB) {
		if (RGB.length === 3) {
			let rgbColor = RGB.map((i) => Math.max(Math.min(i, 255), 0));
			const messageArray = [this.messageIdentifiers.SET_LED_TOGGLE, pulseCount, onTime, offTime, ...rgbColor, 1, 0]
			this.sendMessage(messageArray);
		}
	}

	/******* Internal Helper Functions *******/

	// Send generic message to the die
	sendMessage(messageArray) {
		if (!this.GoDiceService) {
			return Promise.reject(new Error('No Cube characteristic selected yet.'));
		}
		return this.GoDiceService.getCharacteristic('6e400002-b5a3-f393-e0a9-e50e24dcca9e')
			.then(controlPoint => {
				console.debug("messageArray", messageArray);
				const byteMessage = new Uint8Array(messageArray);
				return controlPoint.writeValue(byteMessage).then(response => {
					console.debug("after write response", response);
				});
			});
	}
	
	/**
	* Calculates the closest vector from the table to the given coord
	* @param {dictionary} DieTable - Vector table of current shell
	* @param {dictionary} coord - Vector of die
	*
	**/
	getClosestVector(DieTable, coord) {
		
		const coordX = coord[0]
		const coordY = coord[1]
		const coordZ = coord[2]
				
		var minDistance = Number.MAX_SAFE_INTEGER
		var value = 0
		var vector = []
		var XResult = 0
		var YResult = 0
		var ZResult = 0
		var curDist = 0
		
		// Calculating distance to each value in vector array
		for (let dieValue in DieTable) 
		{	
			vector = DieTable[dieValue]
			
			XResult = coordX - vector[0]
			YResult = coordY - vector[1]
			ZResult = coordZ - vector[2]
			
			// Calculating squared magnitude (since it's only for comparing there's no need for sqrt)
			curDist = ((XResult ** 2) + (YResult ** 2) + (ZResult ** 2))
			
			if (curDist < minDistance) {
				minDistance = curDist
				value = dieValue
			}
			
		}
		return value
	}
	
	// Gets the xyz coord from sent message
	getXyzFromBytes(data, startByte) {
		const x = data.getInt8(startByte);
		const y = data.getInt8(startByte + 1);
		const z = data.getInt8(startByte + 2);
		return [x, y, z]
	}

	// Get Die number from acc raw data according to die type
	getDieValue(data, startByte) {
		
		const xyzArray = this.getXyzFromBytes(data, startByte)
		var value = 0;
		
		// Calculating closest vector according to current die type
		if (this.dieType == GoDice.diceTypes.D6){
			value = this.getClosestVector(GoDice.d6Vectors, xyzArray)
		}
		else if (this.dieType >= GoDice.diceTypes.D20 && this.dieType <= GoDice.diceTypes.D10X) {
			// D20 shell type
			value = this.getClosestVector(GoDice.d20Vectors, xyzArray)
			
			switch (this.dieType) {
				case (GoDice.diceTypes.D10):
					value = GoDice.d10Transform[value]
					break
				case (GoDice.diceTypes.D10X):
					value = GoDice.d10XTransform[value]
					break
			}
		}
		else if (this.dieType >= GoDice.diceTypes.D4 && this.dieType <= GoDice.diceTypes.D12){
			value = this.getClosestVector(GoDice.d24Vectors, xyzArray)
			
			// D24 shell type
			switch (this.dieType) {
				case (GoDice.diceTypes.D4):
					value = GoDice.d4Transform[value]
					break
				case (GoDice.diceTypes.D8):
					value = GoDice.d8Transform[value]
					break
				case (GoDice.diceTypes.D12):
					value = GoDice.d12Transform[value]
					break
			}
		}
		return value

	}
	
	getD20Value (data, startByte) {
		const xyzArray = this.getXyzFromBytes(data, startByte)
		return this.getClosestVector(this.d20Vectors, xyzArray)
	}
	
	getD24Value (data, startByte) {
		const xyzArray = this.getXyzFromBytes(data, startByte)
		return this.getClosestVector(this.d24Vectors, xyzArray)
	}
	
	// Get a message fromn the die and fire the matching event
	parseMessage(data, deviceId) {
		try {
			console.debug("data: ", data);
			console.debug("deviceId: ", deviceId);
			const firstByte = data.getUint8(0);
			if (firstByte === 82) {
				this.onRollStart(deviceId);
				return;
			}

			const secondByte = data.getUint8(1);
			const thirdByte = data.getUint8(2);

			if (firstByte === 66 && secondByte === 97 && thirdByte === 116) {
				this.onBatteryLevel(deviceId, data.getUint8(3));
			}

			if (firstByte === 67 && secondByte === 111 && thirdByte === 108) {
				this.onDiceColor(deviceId, data.getUint8(3));
			}

			if (firstByte === 83) {
				const diceCurrentNumber = this.getDieValue(data, 1);
				const xyzArray = this.getXyzFromBytes(data, 1)
				this.rolledValue = diceCurrentNumber;
				this.onStable(deviceId, diceCurrentNumber, xyzArray);
			}

			if (firstByte === 70 && secondByte === 83) {
				const diceCurrentNumber = this.getDieValue(data, 2);
				const xyzArray = this.getXyzFromBytes(data, 2)
				this.rolledValue = diceCurrentNumber;
				this.onFakeStable(deviceId, diceCurrentNumber, xyzArray);
			}

			if (firstByte === 84 && secondByte === 83) {
				const diceCurrentNumber = this.getDieValue(data, 2);
				const xyzArray = this.getXyzFromBytes(data, 2)
				this.rolledValue = diceCurrentNumber;
				this.onTiltStable(deviceId, xyzArray, diceCurrentNumber);
			}

			if (firstByte === 77 && secondByte === 83) {
				const diceCurrentNumber = this.getDieValue(data, 2);
				const xyzArray = this.getXyzFromBytes(data, 2)
				this.rolledValue = diceCurrentNumber;
				this.onMoveStable(deviceId, diceCurrentNumber, xyzArray);
			}

		}
		catch (err) {
			console.error("err", err);
		}
	}

	/******** Bluetooth Low Energy (BLE) implementation ********/
	handleNotificationChanged(event) {
		this.parseMessage(event.target.value, event.target.service.device.id.toString());
	}

	handleNotificationChanged = this.handleNotificationChanged.bind(this);

	connectDeviceAndCacheCharacteristics() {
		console.debug('Connecting to GATT Server...');
		return this.bluetoothDevice.gatt.connect()
			.then(server => {
				return server.getPrimaryService("6e400001-b5a3-f393-e0a9-e50e24dcca9e");
			})
			.then(service => {
				this.GoDiceService = service;
				return service.getCharacteristic("6e400003-b5a3-f393-e0a9-e50e24dcca9e");
			})
			.then(characteristic => {
				this.CubeCharacteristics = characteristic;
				this.CubeCharacteristics.addEventListener('characteristicvaluechanged', this.handleNotificationChanged);
				return characteristic.getDescriptors();
			})
			.then(descriptors => {
				this.onStartNotificationsButtonClick();
			})
	}

	reConnectDeviceAndCacheCharacteristics() {
		console.debug('Reconnecting to GATT Server...');

		const abortController = new AbortController();

		this.bluetoothDevice.addEventListener('advertisementreceived', (event) => {
			abortController.abort();
		 	this.bluetoothDevice.gatt.connect()
			.then(server => {
				return server.getPrimaryService("6e400001-b5a3-f393-e0a9-e50e24dcca9e");
			})
			.then(service => {
				this.GoDiceService = service;
				return service.getCharacteristic("6e400003-b5a3-f393-e0a9-e50e24dcca9e");
			})
			.then(characteristic => {
				this.CubeCharacteristics = characteristic;
				this.CubeCharacteristics.addEventListener('characteristicvaluechanged', this.handleNotificationChanged);
				return characteristic.getDescriptors();
			})
			.then(descriptors => {
				this.onStartNotificationsButtonClick();
			});
		},{ once: true });

		this.bluetoothDevice.watchAdvertisements({ signal: abortController.signal })
		.catch(error => console.log('Argh! '+ error));
	}


	onStartNotificationsButtonClick() {
		console.debug('Starting Notifications...');
		this.CubeCharacteristics.startNotifications()
			.then(_ => {
				console.debug('onDiceConnected');
				this.onDiceConnected(this.GlobalDeviceId, this);
			})
			.catch(error => {
				console.error('Argh! ' + error);
			});
	}

	onDisconnectButtonClick() {
		if (this.CubeCharacteristics) {
			this.CubeCharacteristics.removeEventListener('characteristicvaluechanged', this.handleNotificationChanged.bind(this));
			this.CubeCharacteristics = null;
		}
		if (this.bluetoothDevice === null)
			return;

		if (this.bluetoothDevice.gatt.connected)
			this.bluetoothDevice.gatt.disconnect();
		else {
			console.debug('> Bluetooth Device is already disconnected');
		}

		// Note that it doesn't disconnect device.
		this.bluetoothDevice = null;
	}

	onDisconnected(event) {
		console.debug('> Bluetooth Device disconnected:' + event);
	}

}


scripts/godice_handler.js
const connectedDice = {};

function LoadStoredInfos(event)
{
	let storedConnectedDice = sessionStorage.getItem('connectedDice');
	if(storedConnectedDice != null)
	{
		document.getElementById("placeholder").textContent= 'Wait... Reloading Stored dices...'
		storedConnectedDice = JSON.parse(storedConnectedDice);	
		Object.values(storedConnectedDice).forEach( function(diceInstance, diceId) {
			newDiceInstance = new GoDice(diceInstance);
			newDiceInstance.reconnectDevice();
		});
	}
	let storedPlayerName = sessionStorage.getItem('playerName');
	document.getElementById('player_name').value = storedPlayerName;
}

if (window.addEventListener) { // Mozilla, Netscape, Firefox
    window.addEventListener('load', LoadStoredInfos, false);
} else if (window.attachEvent) { // IE
    window.attachEvent('onload', LoadStoredInfos);
}

// Open the Bluetooth connection dialog for choosing a GoDice to connect
function openConnectionDialog() {
	const newDice = new GoDice();
	newDice.requestDevice();
}

/**
 * Get a new dice element or it's instance if it already exists
 * @param {string} diceID - the die unique identifier	 
 */
function getDiceHtmlEl(diceID) {
	if (!document.getElementById(diceID)) {
		const newDiceEl = document.createElement("div");
		newDiceEl.id = diceID;
		newDiceEl.className = "dice-wrapper";
		return newDiceEl;
	}
	return document.getElementById(diceID);
}

function selectDiceType(diceInstance)
{
	let diceType = "";
	if(diceInstance.newConnection)
	{	
		//Show popup to select the dice Type
		diceType = prompt("Select the dice Type","D6");
		diceInstance.setDieType(GoDice.diceTypes[diceType]);
	}

	return;
}

function getDiceTypeString(diceInstance)
{
	return Object.keys(GoDice.diceTypes)[diceInstance.getDieType()];
}

function addConnectedDice(diceId, diceInstance)
{
	let diceType = getDiceTypeString(diceInstance);
	diceInstance.setDiceColor();

	placeholder = document.getElementById("placeholder");
	if (placeholder != null) {
       placeholder.remove();
	}

	// get die host from html, where we will put our connected dices
	diceManager = document.getElementById("diceManager");
	diceHtmlEl = getDiceManager(diceId, diceInstance);
	diceManager.appendChild(diceHtmlEl);
	
	diceInstance.getBatteryLevel();

	diceResults = document.getElementById("diceResults");
	diceResult = document.getElementById(diceType+"-group");
	if(diceResult == null)
	{
		diceResult = addDiceResult(diceId, diceInstance);
		diceResults.append(diceResult);
	}

}

GoDice.prototype.onDiceConnected = (diceId, diceInstance) => {

	if(connectedDice[diceId] != null)
	{
		console.log('Dice already connected');
	}
	else
	{
		console.log("Dice connected: ", diceId);
		selectDiceType(diceInstance);
		// Called when a new die is connected - create a dedicate panel for this die
		addConnectedDice(diceId, diceInstance);
		connectedDice[diceId] = diceInstance;
		sessionStorage.setItem('connectedDice', JSON.stringify(connectedDice));
	}
};

GoDice.prototype.onRollStart = (diceId) => {
	console.log("Roll Start: ", diceId);
	let diceType =  getDiceTypeString(connectedDice[diceId]);
	// get rolling indicator
	const diceIndicatorEl = document.getElementById(diceType + "-die-status");
	// show rolling 
	diceIndicatorEl.textContent = "Rolling....";
};

GoDice.prototype.onStable = (diceId, value, xyzArray) => {
	console.log("Stable event: ", diceId, value);
	let diceType =  getDiceTypeString(connectedDice[diceId]);
	// Get roll value indicator and show stable value
	const diceIndicatorEl = document.getElementById(diceType + "-die-status");
	diceIndicatorEl.textContent = "Stable";

	addRoll(diceId, value);
};

GoDice.prototype.onTiltStable = (diceId, xyzArray, value) => {
	console.log("TiltStable: ", diceId, xyzArray);
	let diceType =  getDiceTypeString(connectedDice[diceId]);
	// Get tile indicator and show raw data
	const diceIndicatorEl = document.getElementById(diceType + "-die-status");
	diceIndicatorEl.textContent = "Tilt Stable";

	addRoll(diceId, value);
};

GoDice.prototype.onFakeStable = (diceId, value, xyzArray) => {
	console.log("FakeStable: ", diceId, value);
	let diceType =  getDiceTypeString(connectedDice[diceId]);
	// Get tile indicator and show fake value
	const diceIndicatorEl = document.getElementById(diceType + "-die-status");
	diceIndicatorEl.textContent = "Fake Stable";

	addRoll(diceId, value);
};

GoDice.prototype.onMoveStable = (diceId, value, xyzArray) => {
	console.log("MoveStable: ", diceId, value);
	let diceType =  getDiceTypeString(connectedDice[diceId]);
	// Get tile indicator and show fake value
	const diceIndicatorEl = document.getElementById(diceType + "-die-status");
	diceIndicatorEl.textContent = "Move Stable";

	addRoll(diceId, value);
};

GoDice.prototype.onBatteryLevel = (diceId, batteryLevel) => {
	console.log("BetteryLevel: ", diceId, batteryLevel);
	
	// get dice battery indicator element
	const batteryLevelEl = document.getElementById(diceId + "-battery-indicator");
	connectedDice[diceId].batteryLevel = batteryLevel;
	sessionStorage.setItem('connectedDice', JSON.stringify(connectedDice));

	// put battery level value into battery indicator html element
	batteryLevelEl.textContent = batteryLevel;
	batteryLevelInt = parseInt(batteryLevel);
	if(batteryLevelInt > 80)
		batteryLevelEl.setAttribute('class', 'battery-indicator battery_high');
	else if(batteryLevelInt > 30)
		batteryLevelEl.setAttribute('class', 'battery-indicator battery_medium');
	else
		batteryLevelEl.setAttribute('class', 'battery-indicator battery_low');
};

GoDice.prototype.onDiceColor = (diceId, color) => {
	let colorString =  Object.keys(this.diceColour)[color].toLowerCase();
	console.debug('DiceColor: ', diceId, color, colorString);
	//Set dice Color to class attribute as String
	connectedDice[diceId].dieColor = color;
	sessionStorage.setItem('connectedDice', JSON.stringify(connectedDice));

	const diceColorEl = document.getElementById(diceId+'-type-img');
	if(diceColorEl!= null)
		diceColorEl.style.backgroundColor = colorString;
};

function addRoll(diceId, value)
{
	let diceType =  getDiceTypeString(connectedDice[diceId]);
	let colorString =  Object.keys(connectedDice[diceId].diceColour)[connectedDice[diceId].dieColor].toLowerCase();
	
	removeRolls(diceType, true);

	let diceRolls = document.getElementById(`${diceType}-rolls`);
	let diceRoll = document.createElement('div')
	diceRoll.id = diceType+"roll";
	diceRoll.setAttribute('class','roll');
	diceRoll.textContent = value;
	diceRoll.style.borderColor = colorString;
	diceRolls.appendChild(diceRoll);
}

function removeRolls(diceType, check)
{
	let maxDiceRolls = parseInt(document.getElementById(diceType+'-qty').value,10);
	if(maxDiceRolls == 0)
		maxDiceRolls = 1;

	let diceRolls = document.getElementById(`${diceType}-rolls`);
	if(diceRolls != null && ((check && diceRolls.children.length >= maxDiceRolls) || !check))
	{
		// clear existing die element content if it exists
		while (diceRolls.firstChild) {
	    	diceRolls.removeChild(diceRolls.lastChild);
		}
	}
}


function getDiceManager(diceId, diceInstance)
{
	// get new die element or it's instance if it's already exists
	const diceHtmlEl = getDiceHtmlEl(diceId);
	let diceType =  getDiceTypeString(diceInstance);

	// clear existing die element content if it exists
	while (diceHtmlEl.firstChild) {
        	diceHtmlEl.removeChild(diceHtmlEl.lastChild);
	}

	// add name to die element
	const diceName = document.createElement('div');
	diceName.id = diceId;
	diceName.className = 'dice-name';
	diceName.textContent = `Dice ID: ${diceId}`;
	diceName.setAttribute('hidden', true);
	diceHtmlEl.append(diceName);


	var colorString =  Object.keys(diceInstance.diceColour)[diceInstance.dieColor].toLowerCase();
	diceTypeImg = document.createElement('img');
	diceTypeImg.id = `${diceId}-type-img`;
	diceTypeImg.setAttribute('src', 'images/'+diceType+'.webp');
	diceTypeImg.style.width = '40px';
	diceTypeImg.style.backgroundColor = colorString;
	diceTypeImg.onclick = diceInstance.pulseLed.bind(diceInstance, 5, 30, 20, [0, 0, 255])
	diceHtmlEl.append(diceTypeImg);

	// add battery level indicator
	const batteryIndicator = document.createElement('div');
	batteryIndicator.id = `${diceId}-battery-indicator`;
	batteryIndicator.setAttribute('class','battery-operator');
	batteryIndicator.onclick = diceInstance.getBatteryLevel.bind(diceInstance);
	diceHtmlEl.append(batteryIndicator);

	return diceHtmlEl;
}

function addDiceResult(diceId, diceInstance)
{
		let diceType =  getDiceTypeString(diceInstance);
		var diceResult = document.createElement('div');
		diceResult.id = `${diceType}-group`;
		diceResult.setAttribute('class','diceGroup');

		var diceRollNumber = document.createElement('input');
		diceRollNumber.id = `${diceType}-qty`;
		diceRollNumber.setAttribute('type', 'number');
		diceRollNumber.setAttribute('min', 0);
		diceRollNumber.setAttribute('max', 10);
		diceRollNumber.value = 0;
		diceResult.append(diceRollNumber);

		var diceTypeIndicator = document.createElement('div');
		diceTypeIndicator.id = `${diceType}-label`;
		diceTypeIndicator.setAttribute('class','diceLabel');
		diceTypeIndicator.textContent = diceType.toLowerCase();
		diceResult.append(diceTypeIndicator);
	
		var diceTypeOperator = document.createElement('select');
		diceTypeOperator.id = `${diceType}-operator`;
		diceTypeOperator.setAttribute('class','operator');
		var diceTypeOperOpt = document.createElement('option');
		diceTypeOperOpt.textContent = '+';
		diceTypeOperOpt.value = '+';
		diceTypeOperator.append(diceTypeOperOpt);
		diceTypeOperOpt = document.createElement('option');
		diceTypeOperOpt.textContent = '-';
		diceTypeOperOpt.value = '-';
		diceTypeOperator.append(diceTypeOperOpt);
		diceResult.append(diceTypeOperator);

		var diceTypeBonus = document.createElement('input');
		diceTypeBonus.id = `${diceType}-bonus`;
		diceTypeBonus.setAttribute('type', 'number');
		diceTypeBonus.setAttribute('class', 'bonus');
		diceTypeBonus.setAttribute('min', 0);
		diceTypeBonus.setAttribute('max', 10);
		diceTypeBonus.value = 0;
		diceResult.append(diceTypeBonus);

		var diceStatus = document.createElement('div');
		diceStatus.id = `${diceType}-die-status`;
		diceStatus.setAttribute('class','die-status');
		diceStatus.setAttribute('hidden', true);
		diceResult.append(diceStatus);

		var diceRolls = document.createElement('div');
		diceRolls.id = `${diceType}-rolls`;
		diceRolls.setAttribute('class','rolls');
		diceResult.append(diceRolls);

		var diceCleanRollsEl = document.createElement('button');
		diceCleanRollsEl.id = `${diceType}-clean`;
		diceCleanRollsEl.textContent = "Clean Roll";
		diceCleanRollsEl.setAttribute('class', 'btn btn-outline-primary btn-sm');
		diceCleanRollsEl.onclick = function(){removeRolls(diceType, false)};
		diceResult.append(diceCleanRollsEl);

		return diceResult;
}

function htmlToElement(html) {
    var template = document.createElement('template');
    html = html.trim(); // Never return a text node of whitespace as the result
    template.innerHTML = html;
    return template.content.firstChild;
}

function SendRoll()
{
	url = "https://bembe83.moltenhosting.com/modules/external-dice-roll-connector/roll.html";

	jsonRoll = calculateTotal();
	player = document.getElementById("player").value;

	tab = window.open('about:blank',"_blank");
	tab.document.write(JSON.stringify(jsonRoll));
	tab.document.close();

	$.ajax({
		url: url,
		headers: {
        	'Content-Type': 'application/x-www-form-urlencoded'
    	},
		type: "GET",
		data: { 
		    name: player, 
		    roll: JSON.stringify(jsonRoll)
		  },
		success: function(result){
			tab = window.open('about:blank',"_blank");
			tab.document.write(JSON.stringify(result));
			tab.document.close();	
		},
		error: function(error){
			console.log(error);
		}
	});

}

function calculateTotal()
{
	jsonObj = {};

	total = 0;
	formula = "";
	terms = new Array();
	results = new Array();

	let diceResults = document.getElementById('diceResults');
	let diceGroup = diceResults.querySelectorAll('.diceGroup');

	for (let j = 0; j < diceGroup.length; j++)
	{
		let d = diceGroup[j].id.replace('-group', '').toLowerCase();
		let nrolls = parseInt(document.getElementById(d.toUpperCase()+"-qty").value,10);
		let faces = parseInt(d.replace("d",""),10);
		if(nrolls >0)
		{
			formula = formula + nrolls + d; 
			dice = {};
			dice ["number"] = nrolls;
			dice ["faces"] = faces;

			diceResults = new Array();
			diceTotal = 0
			let diceRolls = diceGroup[j].querySelectorAll('.rolls')[0].querySelectorAll('.roll');

			for (let i = 0; i < nrolls; i++) {
				let roll = 0;
				if(i< diceRolls.length)
	  				roll = parseInt(diceRolls[i].textContent,10);
	  			roll = roll==0?1:roll;

	  			diceResult = {"result": roll,"active": true};
	  			diceResults.push(diceResult);

	  			total = total + roll;
	  			diceTotal = diceTotal +roll;
			}

			dice ["results"] = diceResults;
			terms.push(dice);

			results.push(diceTotal);

			let bonus = parseInt(diceGroup[j].querySelectorAll('.bonus')[0].value, 10);
			let operator = diceGroup[j].querySelectorAll('.operator')[0].selectedOptions[0].value;
			if(operator == "-")
			{
				total = total - bonus;
			}
			else
			{
				operator = "+";
				total = total + bonus;
			}
			if(bonus > 0){
				results.push(operator);
				results.push(bonus);
				formula = formula+operator+bonus+"+";

				jsonOperator = {};
				jsonOperator ["operator"] = operator;
				jsonNumber = {};
				jsonNumber ["number"] = bonus
				terms.push(jsonOperator);
				terms.push(jsonNumber);
			}
			results.push("+");
		}
	}
	results.pop();
	jsonObj ["formula"] = formula.slice(0,-1);
	jsonObj ["terms"] = terms;
	jsonObj ["results"] = results;
	jsonObj ["_total"] = total;


	return jsonObj;
}

function storePlayerName()
{
	let storedPlayerName = document.getElementById('player_name').value;
	 sessionStorage.setItem('playerName',storedPlayerName);
}

